### PR TITLE
Fix tables width, paginators, button on mobile / touchscreens

### PR DESCRIPTION
**Pull request by kimsible**
Proposition for the issue : https://github.com/Chocobozzz/PeerTube/issues/2694

It also fixes the display of actions buttons in an `overflow` table view.

**Before**
my-account-tables-1

**After**
my-account-tables-1-fix

### CODE REVIEW

File: client/src/app/+my-account/my-account.component.scss [hunk 1]
```diff
@@ -8,3 +8,63 @@
...
+              margin-left: 5px;
+            }
+
+            :first-child {
+              margin-left: 0px;
+            }
+          }
+        }
+      }
+    }
```
Comment (rigelk) [inline client/src/app/+my-account/my-account.component.scss:38]: This block might be factorized in a mixin.

**Comment by rigelk** [comment]
Having a table that displays properly on mobile is something that we might want to apply not just to that view, but also to all instances of `p-table` used throughout PeerTube. Your fix is generic enough to improve that situtation on other tables by placing it in `application.scss`.

It is a ton of work to check all other tables for regressions, but the potential value is huge for mobile users. Potentially in another PR or by someone else, of course :slightly_smiling_face: What do you think?

**Comment by kimsible** [comment]
> Having a table that displays properly on mobile is something that we might want to apply not just to that view, but also to all instances of `p-table` used throughout PeerTube. Your fix is generic enough to improve that situtation on other tables by placing it in `application.scss`.


Thanks for the review, I first thought about all tables, but admins have `dropdown` menus (so it will rase the same issue than touchscreens on the playlist :/). 

> It is a ton of work to check all other tables for regressions, but the potential value is huge for mobile users. Potentially in another PR or by someone else, of course slightly_smiling_face What do you think?

Yes ! I agree, plus it will be very useful for moderators to have an access on touchscreens.

I can place that code in  `application.scss` and exclude admin tables ?

**Comment by rigelk** [comment]
> Thanks for the review, I first thought about all tables, but admins have `dropdown` menus (so it will raise the same issue than touchscreens on the playlist :/).

You are only affecting the container's `overflow-x` this time, so that problem will not repeat itself there. Probably. :sweat_smile: 

> I can place that code in `application.scss` and exclude admin tables ?

Maybe more `primeng-custom.scss` since the change mainly affects `p-table`. You can do a first version excluding the admin tables if you fear regressions there, but you should test visually any other affected table - because that's what I will have to do anyway :wink: 

> I forgot to change the `:hover` event to a continuously display for touchscreens

What do you mean?

**Comment by kimsible** [comment]
> > Thanks for the review, I first thought about all tables, but admins have `dropdown` menus (so it will raise the same issue than touchscreens on the playlist :/).
> 
> You are only affecting the container's `overflow-x` this time, so that problem will not repeat itself there. Probably. sweat_smile
> 
With the same trick of the user dropdown on the left menu I suppose ?

> > I forgot to change the `:hover` event to a continuously display for touchscreens
> 
> What do you mean?

Action-buttons appear on `td:hover` this event does not exist on touchscreen, no ?

**Comment by rigelk** [comment]
> > You are only affecting the container's `overflow-x` this time, so that problem will not repeat itself there. Probably. sweat_smile
> 
> With the same trick of the user dropdown on the left menu I suppose ?

No, I mean that the dropdown doesn't have the issue currently (no overflow rule present) and won't have it with your additions since you only affect the horizontal overflow, and the dropdown… well… drops vertically. Am I overlooking something?

> Action-buttons appear on the line hover event `tr:hover` this event does not exist on touchscreen, no ?

Ah, for action buttons. You're right.

**Comment by kimsible** [comment]
These last commits apply this on every tables (including admin), also :

- Keep displaying and fix action-cell elements on touchscreens (not :hover)

my-account-tables-ownership-mobile


- Fix paginator on mobile views

admin-tables-showing
admin-tables-showing-fix-2



- Apply dropdown trick the `dropdown-root` on action-cell 

admin-tables-dropdown
admin-tables-dropdown-fix

**Comment by kimsible** [comment]
@rigelk I fixed the `min-width` of the table to `800px / $small-view`, do you think it is enough for admins tables ?

**Comment by rigelk** [comment]
> @rigelk I fixed the `min-width` of the table to `800px / $small-view`, do you think it is enough for admins tables ?

It might be a bit short. 900px/lg breakpoint is better.

**Comment by kimsible** [comment]
@rigelk thanks, just to inform that this PR may depend on that one about contents responsive styles (expanded/not-expanded) : https://github.com/Chocobozzz/PeerTube/pull/2698

File: client/src/sass/application.scss [hunk 2]
```diff
@@ -331,3 +331,38 @@ table {
...
+@media screen and (max-width: #{map-get($grid-breakpoints, lg)}) {
+  .main-col {
+    &.expanded {
+      @include overflow-datatable(map-get($grid-breakpoints, lg), $expanded-horizontal-margins/3);
+    }
+
+    &:not(.expanded) {
+      // TODO: $horizontal-margins might change to $expanded-horizontal-margins/3 + $menu-width/2
+      @include overflow-datatable(map-get($grid-breakpoints, lg), $not-expanded-horizontal-margins + $menu-width/2);
+    }
```
Comment (kimsible) [inline client/src/sass/application.scss:358]: Required the PR https://github.com/Chocobozzz/PeerTube/pull/2698 to change the `$horizontal-margins` and compact this block :

Comment (rigelk) [reply client/src/sass/application.scss:358]: I merged the aforementioned PR.

File: client/src/app/shared/buttons/action-dropdown.component.html [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<div class="dropdown-root" ngbDropdown [placement]="placement" *ngIf="areActionsDisplayed(actions, entry)">
+<div class="dropdown-root" ngbDropdown [placement]="placement" container="body" autoClose="outside" *ngIf="areActionsDisplayed(actions, entry)">
```
Comment (rigelk) [inline client/src/app/shared/buttons/action-dropdown.component.html:1]:
```suggestion
<div class="dropdown-root" ngbDropdown [placement]="placement" container="body" *ngIf="areActionsDisplayed(actions, entry)">
```

You cannot use `autoClose="outside"`: the action-dropdown then doesn't close on menu clicks, which is required in some scenarios (coming to mind, the addition of a video in a playlist, where the previous dropdown is kept open as the new one opens…).

File: client/src/sass/application.scss (hunk 2 above)
Comment (rigelk) [reply client/src/sass/application.scss:358]: It got merged :slightly_smiling_face:

**Comment by kimsible** [comment]
Thanks @rigelk, I think about the paginator on mobile, the items are well limited to 5 ?

**Comment by kimsible** [comment]
> @kimsible what de you mean?

More we have items more we have pages in the paginator (seems to be limited to 5) but there is a solution using the "Programmatic Control" https://www.primefaces.org/primeng/showcase/#/table/page

**Comment by rigelk** [comment]
> More we have items more we have pages in the paginator (seems to be limited to 5) but there is a solution using the "Programmatic Control" [primefaces.org/primeng/showcase/#/table/page](https://www.primefaces.org/primeng/showcase/#/table/page)

IMHO simpler: use `pageLinkSize="n"` on the `p-paginator` component, where `n` is the number of pages we wish to display.

**Comment by kimsible** [comment]
> > More we have items more we have pages in the paginator (seems to be limited to 5) but there is a solution using the "Programmatic Control" [primefaces.org/primeng/showcase/#/table/page](https://www.primefaces.org/primeng/showcase/#/table/page)
> 
> IMHO simpler: use `pageLinkSize="n"` on the `p-paginator` component, where `n` is the number of pages we wish to display.
> 
> EDIT: c.f. https://www.primefaces.org/primeng/showcase/#/paginator

Thanks, maybe the Programmatic Control in another PR later ?

**Comment by rigelk** [comment]
> Thanks, maybe the Programmatic Control in another PR later ?

Isn't limiting the paginator via `pageLinkSize` enough? At size 1 it gives the same behaviour and is one component less to maintain :slightly_smiling_face:

**Comment by kimsible** [comment]
> Isn't limiting the paginator via `pageLinkSize` enough? At size 1 it gives the same behaviour and is one component less to maintain slightly_smiling_face

size 1 is enough, we need the screenService on each table to change dynamically the pageLinkSize

**Comment by rigelk** [comment]
> size 1 is enough, we need the screenService on each table to change dynamically the pageLinkSize

It would be more manageable in another PR.

Could you resolve conflicts and rebase on develop first?

**Comment by kimsible** [comment]
> > size 1 is enough, we need the screenService on each table to change dynamically the pageLinkSize
> 
> It would be more manageable in another PR.

Yes it seems not so easy to implement..., btw, 5 pages-links appears on a screen width ~400px.

> Could you resolve conflicts and rebase on develop first?

I did it on the last commit

**Comment by kimsible** [comment]
> LGTM! Thanks so much for taking the time to go the extra mile to make the table improvement generic smiley

Yeah ! next time the paginator, you're welcome, using PeerTube on tablets and big smartphones is very asked :)

File: client/src/sass/application.scss [hunk 4]
```diff
@@ -364,51 +363,16 @@ table {
     }
 
     &:not(.expanded) {
-      // TODO: $horizontal-margins might change to $expanded-horizontal-margins/3 + $menu-width/2
       @include overflow-datatable(breakpoint(lg), $not-expanded-horizontal-margins + $menu-width/2);
```
Comment (kimsible) [inline client/src/sass/application.scss:366]: @rigelk when you rebased you forgot to change the `$horizontal-margins` to `$expanded-horizontal-margins/3` here, no ?

Comment (rigelk) [reply client/src/sass/application.scss:366]: Ah, right. I assumed you had forgotten to remove the TODO since your PR wasn't a draft anymore.

Comment (rigelk) [reply client/src/sass/application.scss:366]: I'll change that in the next commit I do.

Comment (kimsible) [reply client/src/sass/application.scss:366]: Thanks !

Comment (rigelk) [reply client/src/sass/application.scss:366]: @kimsible I integrated the change in c2a89b70c

**Pull request merged**